File: games/seven_wonders_duel/bots.py
"""Baseline players and reproducible match runners."""

from __future__ import annotations

from dataclasses import dataclass
import random
from typing import Protocol

from .data import CARDS_BY_NAME, PROGRESS_BY_NAME, WONDERS_BY_NAME, CardColor
from .engine import Action, ActionUse, apply_action, legal_actions, score_player
from .game import GameState, Phase, VictoryType, new_game
from .rules import Resource


class Bot(Protocol):
    name: str

    def select_action(self, game: GameState) -> Action: ...


def _require_actions(game: GameState) -> tuple[Action, ...]:
    actions = legal_actions(game)
    if not actions:
        raise ValueError("no legal action is available")
    return actions


class RandomBot:
    name = "random"

    def __init__(self, seed: int = 0):
        self._rng = random.Random(seed)

    def select_action(self, game: GameState) -> Action:
        return self._rng.choice(_require_actions(game))


def _science_count(game: GameState, player: int) -> int:
    return len(_science_symbols(game, player))


def _science_symbols(game: GameState, player: int) -> frozenset:
    symbols = {
        CARDS_BY_NAME[name].science
        for name in game.cities[player].buildings
        if CARDS_BY_NAME[name].science is not None
    }
    symbols.update(
        PROGRESS_BY_NAME[name].science
        for name in game.cities[player].progress_tokens
        if PROGRESS_BY_NAME[name].science is not None
    )
    return frozenset(symbols)


def _actor(game: GameState) -> int:
    return (
        game.pending_choice.player
        if game.pending_choice is not None
        else game.active_player
    )


def _economy_value(game: GameState, player: int) -> float:
    value = 0.0
    for name in game.cities[player].buildings:
        card = CARDS_BY_NAME[name]
        value += 0.7 * len(card.fixed_production)
        if card.choice_production:
            value += 0.8 + 0.2 * (len(card.choice_production) - 1)
        value += 0.7 * len(card.trade_discount)
    for name in game.cities[player].built_wonders:
        choices = WONDERS_BY_NAME[name].choice_production
        if choices:
            value += 0.8 + 0.2 * (len(choices) - 1)
    return value


def _unbuilt_wonder_value(game: GameState, player: int) -> float:
    city = game.cities[player]
    value = 0.0
    for name in city.wonders:
        if name in city.built_wonders or name in game.retired_wonders:
            continue
        wonder = WONDERS_BY_NAME[name]
        value += 0.25 * wonder.victory_points + 0.75 * wonder.shields
        if wonder.choice_production:
            value += 0.6
        for effect in wonder.effects:
            if effect.kind.value == "play_again":
                value += 1.2
            elif effect.kind.value in {
                "destroy_opponent_brown",
                "destroy_opponent_grey",
                "build_from_discard_free",
                "choose_unused_progress",
            }:
                value += 0.8
            elif effect.kind.value == "immediate_coins":
                value += 0.06 * effect.amount
    return value


def evaluate_state(game: GameState, player: int) -> float:
    """Public-feature heuristic from ``player``'s perspective."""

    opponent = 1 - player
    if game.phase is Phase.COMPLETE:
        if game.winner == player:
            return 1_000_000.0
        if game.winner == opponent:
            return -1_000_000.0
        return 0.0

    scores = (score_player(game, player), score_player(game, opponent))
    military = game.conflict_position if player == 0 else -game.conflict_position
    value = 8.0 * (scores[0].total - scores[1].total)
    value += 0.2 * (game.cities[player].coins - game.cities[opponent].coins)
    value += 2.0 * military
    value += 4.0 * (_science_count(game, player) - _science_count(game, opponent))
    value += 1.2 * (_economy_value(game, player) - _economy_value(game, opponent))
    value += 0.5 * (
        _unbuilt_wonder_value(game, player) - _unbuilt_wonder_value(game, opponent)
    )
    if game.active_player == player:
        value += 0.1
    return value


def _action_tiebreak(action: Action) -> tuple:
    return (
        action.use.value,
        action.slot_id or (-1, -1),
        action.wonder_name or "",
        action.choice or "",
        -1 if action.starting_player is None else action.starting_player,
    )


class GreedyBot:
    """Deterministic one-ply baseline using only public state features."""

    name = "greedy"

    def select_action(self, game: GameState) -> Action:
        player = _actor(game)
        actions = _require_actions(game)
        scored: list[tuple[float, tuple, Action]] = []
        for action in actions:
            child = game.clone()
            apply_action(child, action)
            scored.append(
                (evaluate_state(child, player), _action_tiebreak(action), action)
            )
        return max(scored, key=lambda entry: (entry[0], entry[1]))[2]


# Rush bots are an early-training curriculum, not general-purpose evaluators.
# They intentionally overvalue one supremacy route so successful examples make
# science and military pressure visible to the policy/value heads immediately.

_SCIENCE_RESOURCE_WEIGHTS = {
    Resource.WOOD: 1.1,
    Resource.CLAY: 0.8,
    Resource.STONE: 1.0,
    Resource.GLASS: 1.5,
    Resource.PAPYRUS: 1.5,
}
_MILITARY_RESOURCE_WEIGHTS = {
    Resource.WOOD: 1.4,
    Resource.CLAY: 1.4,
    Resource.STONE: 1.4,
    Resource.GLASS: 0.7,
    Resource.PAPYRUS: 0.8,
}

_SCIENCE_WONDER_VALUES = {
    "The Great Library": 420.0,
    "The Mausoleum": 300.0,
    "Piraeus": 170.0,
    "The Great Lighthouse": 150.0,
    "The Hanging Gardens": 130.0,
    "The Sphinx": 130.0,
    "The Temple of Artemis": 120.0,
    "The Appian Way": 100.0,
}
_MILITARY_WONDER_VALUES = {
    "The Colossus": 440.0,
    "The Statue of Zeus": 360.0,
    "Circus Maximus": 340.0,
    "The Appian Way": 160.0,
    "The Hanging Gardens": 130.0,
    "The Sphinx": 130.0,
    "Piraeus": 120.0,
    "The Temple of Artemis": 110.0,
}


def _card_for_action(game: GameState, action: Action):
    if action.slot_id is None:
        return None
    card = game.tableau.cards.get(action.slot_id)
    if card is None or not card.present or not card.revealed:
        return None
    return CARDS_BY_NAME[card.card_name]


def _resource_support_value(
    game: GameState, player: int, weights: dict[Resource, float]
) -> float:
    """Public economic support for a rush route.

    Flexible producers count as one unit of their best supported resource.
    This deliberately values capability rather than peeking at future cards.
    """

    value = 0.0
    for name in game.cities[player].buildings:
        card = CARDS_BY_NAME[name]
        value += sum(weights[resource] for resource in card.fixed_production)
        if card.choice_production:
            value += max(weights[resource] for resource in card.choice_production)
        value += 0.8 * sum(weights[resource] for resource in card.trade_discount)
    for name in game.cities[player].built_wonders:
        choices = WONDERS_BY_NAME[name].choice_production
        if choices:
            value += max(weights[resource] for resource in choices)
    return value


def _progress_is_off_board(game: GameState, token_name: str) -> bool:
    if token_name in game.available_progress_tokens:
        return False
    return not any(
        token_name in city.progress_tokens for city in game.cities
    )


def _relative_military(game: GameState, player: int) -> int:
    return game.conflict_position if player == 0 else -game.conflict_position


def _military_pressure(position: int) -> float:
    # Odd, monotonic, and convex toward either capital.
    return 25.0 * position + 8.0 * position * abs(position)


class _RushBot:
    """Shared reproducible one-ply shell for the four curriculum bots.

    Action scoring reads the action identity and public city/track state only.
    Applying an action to a clone may resolve a reveal, but no score reads the
    revealed tableau identity, so the policy is invariant to hidden deals.
    """

    name = "rush"
    target_victory: VictoryType

    def __init__(
        self,
        seed: int = 0,
        *,
        exploration: float = 0.0,
        top_choices: int = 3,
    ):
        if not 0.0 <= exploration <= 1.0:
            raise ValueError("exploration must be in [0, 1]")
        if top_choices <= 0:
            raise ValueError("top_choices must be positive")
        self._rng = random.Random(seed)
        self.exploration = exploration
        self.top_choices = top_choices

    def select_action(self, game: GameState) -> Action:
        player = _actor(game)
        scored: list[tuple[float, tuple, Action]] = []
        for action in _require_actions(game):
            score = self._score_action(game, action, player)
            scored.append((score, _action_tiebreak(action), action))
        scored.sort(key=lambda entry: (entry[0], entry[1]), reverse=True)
        if self.exploration and self._rng.random() < self.exploration:
            choices = scored[: min(self.top_choices, len(scored))]
            return self._rng.choice(choices)[2]
        return scored[0][2]

    def _score_action(self, game: GameState, action: Action, player: int) -> float:
        card = _card_for_action(game, action)
        child = game.clone()
        apply_action(child, action)

        if child.phase is Phase.COMPLETE:
            if child.winner == player:
                return (
                    3_000_000.0
                    if child.victory_type is self.target_victory
                    else 2_000_000.0
                )
            if child.winner is None:
                return 0.0
            return -3_000_000.0

        score = 0.02 * evaluate_state(child, player)
        score += self._focus_score(game, child, action, player, card)
        if _actor(child) == player:
            score += 25.0  # replay Wonders and pending choices are real tempo
        if (
            action.use is ActionUse.CHOOSE_NEXT_START_PLAYER
            and action.starting_player == player
        ):
            score += 15.0
        return score

    def _focus_score(self, game, child, action, player, card) -> float:
        raise NotImplementedError


class _ScienceRushBot(_RushBot):
    target_victory = VictoryType.SCIENTIFIC
    economy_weight = 8.0
    coin_weight = 0.4
    economy_first = False

    def _focus_score(self, game, child, action, player, card) -> float:
        before_symbols = _science_symbols(game, player)
        after_symbols = _science_symbols(child, player)
        new_symbols = len(after_symbols - before_symbols)
        before_pairs = len(game.cities[player].claimed_science_pairs)
        after_pairs = len(child.cities[player].claimed_science_pairs)
        pair_delta = after_pairs - before_pairs

        score = 0.0
        if new_symbols:
            score += new_symbols * (700.0 + 140.0 * len(before_symbols))
        if pair_delta:
            law_live = (
                "Law" in game.available_progress_tokens
                and "Law" not in game.cities[player].progress_tokens
            )
            score += pair_delta * (900.0 if law_live else 260.0)

        if card is not None and card.color is CardColor.GREEN:
            if action.use is ActionUse.CONSTRUCT_BUILDING:
                score += 100.0
                if card.age == 1 and card.science not in before_symbols:
                    score += 1_100.0  # hard Age-I symbol-family override
            elif action.use in (
                ActionUse.DISCARD_FOR_COINS,
                ActionUse.CONSTRUCT_WONDER,
            ):
                score -= 900.0 if card.age == 1 else 350.0

        wonder_name = action.wonder_name
        if wonder_name is not None:
            score += _SCIENCE_WONDER_VALUES.get(wonder_name, 0.0)
            if wonder_name == "The Great Library" and _progress_is_off_board(
                game, "Law"
            ):
                score += 280.0
            if wonder_name == "The Mausoleum":
                score += 80.0 * sum(
                    CARDS_BY_NAME[name].color is CardColor.GREEN
                    for name in game.discard_pile
                )
            if self.economy_first and wonder_name in {
                "Piraeus",
                "The Great Lighthouse",
            }:
                score += 130.0

        if action.use is ActionUse.RESOLVE_PENDING_CHOICE and action.choice:
            if action.choice == "Law":
                score += 1_200.0
            elif action.choice in {"Urbanism", "Agriculture"}:
                score += 180.0
            elif action.choice == "Economy":
                score += 100.0

        support_delta = _resource_support_value(
            child, player, _SCIENCE_RESOURCE_WEIGHTS
        ) - _resource_support_value(game, player, _SCIENCE_RESOURCE_WEIGHTS)
        coin_delta = child.cities[player].coins - game.cities[player].coins
        score += self.economy_weight * support_delta + self.coin_weight * coin_delta
        return score


class ScienceAggressiveBot(_ScienceRushBot):
    """Takes missing science symbols immediately and forces the Law/Library route."""

    name = "science_aggressive/v1"


class ScienceEconomyBot(_ScienceRushBot):
    """Builds science-enabling production between immediate symbol opportunities."""

    name = "science_economy/v1"
    economy_weight = 55.0
    coin_weight = 1.8
    economy_first = True


class _MilitaryRushBot(_RushBot):
    target_victory = VictoryType.MILITARY
    economy_weight = 8.0
    coin_weight = 0.3
    economy_first = False

    def _focus_score(self, game, child, action, player, card) -> float:
        before_position = _relative_military(game, player)
        after_position = _relative_military(child, player)
        score = 3.0 * (
            _military_pressure(after_position) - _military_pressure(before_position)
        )

        if card is not None and card.color is CardColor.RED:
            if action.use is ActionUse.CONSTRUCT_BUILDING:
                score += 650.0 if card.age == 1 else 300.0
            elif action.use in (
                ActionUse.DISCARD_FOR_COINS,
                ActionUse.CONSTRUCT_WONDER,
            ):
                score -= 650.0 if card.age == 1 else 300.0

        wonder_name = action.wonder_name
        if wonder_name is not None:
            score += _MILITARY_WONDER_VALUES.get(wonder_name, 0.0)
            if self.economy_first and wonder_name in {
                "Piraeus",
                "The Great Lighthouse",
            }:
                score += 90.0

        if action.use is ActionUse.RESOLVE_PENDING_CHOICE and action.choice:
            if action.choice == "Strategy":
                score += 1_200.0
            elif action.choice in {"Urbanism", "Agriculture"}:
                score += 170.0
            elif action.choice == "Economy":
                score += 100.0

        opponent = 1 - player
        coin_damage = game.cities[opponent].coins - child.cities[opponent].coins
        score += 8.0 * max(0, coin_damage)
        support_delta = _resource_support_value(
            child, player, _MILITARY_RESOURCE_WEIGHTS
        ) - _resource_support_value(game, player, _MILITARY_RESOURCE_WEIGHTS)
        coin_delta = child.cities[player].coins - game.cities[player].coins
        score += self.economy_weight * support_delta + self.coin_weight * coin_delta
        return score


class MilitaryAggressiveBot(_MilitaryRushBot):
    """Builds shields immediately and values track pressure increasingly near 9."""

    name = "military_aggressive/v1"


class MilitaryEconomyBot(_MilitaryRushBot):
    """Builds military-enabling production between immediate red-card chances."""

    name = "military_economy/v1"
    economy_weight = 55.0
    coin_weight = 1.6
    economy_first = True


@dataclass(frozen=True, slots=True)
class MatchResult:
    seed: int
    winner: int | None
    victory_type: VictoryType
    final_scores: tuple[int, int] | None
    actions: int


@dataclass(frozen=True, slots=True)
class SeriesResult:
    games: int
    bot_a_wins: int
    bot_b_wins: int
    draws: int
    military: int
    scientific: int
    civilian: int
    average_actions: float


def play_game(
    player_zero: Bot,
    player_one: Bot,
    *,
    seed: int,
    first_player: int = 0,
    max_actions: int = 256,
) -> MatchResult:
    game = new_game(seed=seed, first_player=first_player)
    bots = (player_zero, player_one)
    action_count = 0
    while game.phase is not Phase.COMPLETE:
        if action_count >= max_actions:
            raise RuntimeError(f"game exceeded {max_actions} actions")
        player = _actor(game)
        action = bots[player].select_action(game)
        if action not in legal_actions(game):
            raise ValueError(f"{bots[player].name} returned illegal action: {action}")
        apply_action(game, action)
        action_count += 1
    if game.victory_type is None:
        raise AssertionError("terminal game is missing a victory type")
    return MatchResult(
        seed=seed,
        winner=game.winner,
        victory_type=game.victory_type,
        final_scores=game.final_scores,
        actions=action_count,
    )


def play_series(
    bot_a: Bot,
    bot_b: Bot,
    *,
    games: int,
    seed: int = 0,
) -> SeriesResult:
    if games <= 0:
        raise ValueError("games must be positive")
    a_wins = b_wins = draws = 0
    victory_counts = {kind: 0 for kind in VictoryType}
    total_actions = 0
    for index in range(games):
        a_is_zero = index % 2 == 0
        result = play_game(
            bot_a if a_is_zero else bot_b,
            bot_b if a_is_zero else bot_a,
            seed=seed + index,
            first_player=0,
        )
        total_actions += result.actions
        victory_counts[result.victory_type] += 1
        if result.winner is None:
            draws += 1
        elif (result.winner == 0) == a_is_zero:
            a_wins += 1
        else:
            b_wins += 1
    return SeriesResult(
        games=games,
        bot_a_wins=a_wins,
        bot_b_wins=b_wins,
        draws=draws,
        military=victory_counts[VictoryType.MILITARY],
        scientific=victory_counts[VictoryType.SCIENTIFIC],
        civilian=(
            victory_counts[VictoryType.CIVILIAN]
            + victory_counts[VictoryType.SHARED_CIVILIAN]
        ),
        average_actions=total_actions / games,
    )

File: games/seven_wonders_duel/seven_wonders_rust/src/bots.rs
//! Rust ports of the Phase-D curriculum and anchor bots.

use crate::codec::encode_action;
use crate::data::{self, CardColor, EffectKind, Resource};
use crate::engine::{Action, ActionUse};
use crate::rng::Rng;
use crate::state::{GameState, PendingChoiceKind, Phase, VictoryType};
use std::cmp::Ordering;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum BotKind {
    Greedy,
    ScienceAggressive,
    ScienceEconomy,
    MilitaryAggressive,
    MilitaryEconomy,
}

impl BotKind {
    pub fn parse(name: &str) -> Option<Self> {
        match name {
            "greedy" => Some(Self::Greedy),
            "science_aggressive/v1" => Some(Self::ScienceAggressive),
            "science_economy/v1" => Some(Self::ScienceEconomy),
            "military_aggressive/v1" => Some(Self::MilitaryAggressive),
            "military_economy/v1" => Some(Self::MilitaryEconomy),
            _ => None,
        }
    }

    pub fn name(self) -> &'static str {
        match self {
            Self::Greedy => "greedy",
            Self::ScienceAggressive => "science_aggressive/v1",
            Self::ScienceEconomy => "science_economy/v1",
            Self::MilitaryAggressive => "military_aggressive/v1",
            Self::MilitaryEconomy => "military_economy/v1",
        }
    }
}

fn actor(g: &GameState) -> usize {
    g.pending_choice
        .as_ref()
        .map_or(g.active_player, |choice| choice.player)
}

fn science_symbols(g: &GameState, player: usize) -> [bool; 7] {
    let mut symbols = [false; 7];
    for &cid in &g.cities[player].buildings {
        if let Some(symbol) = data::card(cid).science {
            symbols[symbol as usize] = true;
        }
    }
    for &pid in &g.cities[player].progress_tokens {
        if let Some(symbol) = data::progress(pid).science {
            symbols[symbol as usize] = true;
        }
    }
    symbols
}

fn science_count(g: &GameState, player: usize) -> usize {
    science_symbols(g, player)
        .into_iter()
        .filter(|&x| x)
        .count()
}

fn economy_value(g: &GameState, player: usize) -> f64 {
    let mut value = 0.0;
    for &cid in &g.cities[player].buildings {
        let card = data::card(cid);
        value += 0.7 * card.fixed_production.len() as f64;
        if !card.choice_production.is_empty() {
            value += 0.8 + 0.2 * (card.choice_production.len() - 1) as f64;
        }
        value += 0.7 * card.trade_discount.len() as f64;
    }
    for &wid in &g.cities[player].built_wonders {
        let choices = data::wonder(wid).choice_production;
        if !choices.is_empty() {
            value += 0.8 + 0.2 * (choices.len() - 1) as f64;
        }
    }
    value
}

fn unbuilt_wonder_value(g: &GameState, player: usize) -> f64 {
    let city = &g.cities[player];
    let mut value = 0.0;
    for &wid in &city.wonders {
        if city.built_wonders.contains(&wid) || g.retired_wonders.contains(&wid) {
            continue;
        }
        let wonder = data::wonder(wid);
        value += 0.25 * wonder.victory_points as f64 + 0.75 * wonder.shields as f64;
        if !wonder.choice_production.is_empty() {
            value += 0.6;
        }
        for effect in wonder.effects {
            value += match effect.kind {
                EffectKind::PlayAgain => 1.2,
                EffectKind::DestroyOpponentBrown
                | EffectKind::DestroyOpponentGrey
                | EffectKind::BuildFromDiscardFree
                | EffectKind::ChooseUnusedProgress => 0.8,
                EffectKind::ImmediateCoins => 0.06 * effect.amount as f64,
                _ => 0.0,
            };
        }
    }
    value
}

fn evaluate_state(g: &GameState, player: usize) -> f64 {
    let opponent = 1 - player;
    if g.phase == Phase::Complete {
        return if g.winner == Some(player) {
            1_000_000.0
        } else if g.winner == Some(opponent) {
            -1_000_000.0
        } else {
            0.0
        };
    }
    let scores = (g.score_player(player).total, g.score_player(opponent).total);
    let military = if player == 0 {
        g.conflict_position
    } else {
        -g.conflict_position
    };
    let mut value = 8.0 * (scores.0 - scores.1) as f64;
    value += 0.2 * (g.cities[player].coins - g.cities[opponent].coins) as f64;
    value += 2.0 * military as f64;
    value += 4.0 * (science_count(g, player) as i32 - science_count(g, opponent) as i32) as f64;
    value += 1.2 * (economy_value(g, player) - economy_value(g, opponent));
    value += 0.5 * (unbuilt_wonder_value(g, player) - unbuilt_wonder_value(g, opponent));
    if g.active_player == player {
        value += 0.1;
    }
    value
}

fn use_rank(use_: ActionUse) -> u8 {
    match use_ {
        ActionUse::ChooseNextStartPlayer => 0,
        ActionUse::ConstructBuilding => 1,
        ActionUse::ConstructWonder => 2,
        ActionUse::DiscardForCoins => 3,
        ActionUse::DraftWonder => 4,
        ActionUse::ResolvePendingChoice => 5,
    }
}

fn choice_name(g: &GameState, action: &Action) -> &'static str {
    let Some(choice) = action.choice else {
        return "";
    };
    match g.pending_choice.as_ref().map(|pending| pending.kind) {
        Some(PendingChoiceKind::DestroyOpponentBrown)
        | Some(PendingChoiceKind::DestroyOpponentGrey)
        | Some(PendingChoiceKind::BuildFromDiscardFree) => data::card(choice).name,
        Some(PendingChoiceKind::ChooseUnusedProgress)
        | Some(PendingChoiceKind::ChooseAvailableProgress) => data::progress(choice).name,
        None => "",
    }
}

fn action_key(g: &GameState, action: &Action) -> (u8, (i32, i32), &'static str, &'static str, i32) {
    (
        use_rank(action.use_),
        action.slot.map_or((-1, -1), |slot| g.tableau.slot_id(slot)),
        action.wonder.map_or("", |wid| data::wonder(wid).name),
        choice_name(g, action),
        action.starting_player.map_or(-1, |seat| seat as i32),
    )
}

fn resource_weight(resource: Resource, science: bool) -> f64 {
    match (science, resource) {
        (true, Resource::Wood) => 1.1,
        (true, Resource::Clay) => 0.8,
        (true, Resource::Stone) => 1.0,
        (true, Resource::Glass | Resource::Papyrus) => 1.5,
        (false, Resource::Wood | Resource::Clay | Resource::Stone) => 1.4,
        (false, Resource::Glass) => 0.7,
        (false, Resource::Papyrus) => 0.8,
    }
}

fn resource_support_value(g: &GameState, player: usize, science: bool) -> f64 {
    let mut value = 0.0;
    for &cid in &g.cities[player].buildings {
        let card = data::card(cid);
        value += card
            .fixed_production
            .iter()
            .map(|&r| resource_weight(r, science))
            .sum::<f64>();
        if !card.choice_production.is_empty() {
            value += card
                .choice_production
                .iter()
                .map(|&r| resource_weight(r, science))
                .fold(f64::NEG_INFINITY, f64::max);
        }
        value += 0.8
            * card
                .trade_discount
                .iter()
                .map(|&r| resource_weight(r, science))
                .sum::<f64>();
    }
    for &wid in &g.cities[player].built_wonders {
        let choices = data::wonder(wid).choice_production;
        if !choices.is_empty() {
            value += choices
                .iter()
                .map(|&r| resource_weight(r, science))
                .fold(f64::NEG_INFINITY, f64::max);
        }
    }
    value
}

fn progress_is_off_board(g: &GameState, name: &str) -> bool {
    let pid = data::progress_id(name);
    !g.available_progress_tokens.contains(&pid)
        && !g
            .cities
            .iter()
            .any(|city| city.progress_tokens.contains(&pid))
}

fn wonder_value(name: &str, science: bool) -> f64 {
    match (science, name) {
        (true, "The Great Library") => 420.0,
        (true, "The Mausoleum") => 300.0,
        (true, "Piraeus") => 170.0,
        (true, "The Great Lighthouse") => 150.0,
        (true, "The Hanging Gardens" | "The Sphinx") => 130.0,
        (true, "The Temple of Artemis") => 120.0,
        (true, "The Appian Way") => 100.0,
        (false, "The Colossus") => 440.0,
        (false, "The Statue of Zeus") => 360.0,
        (false, "Circus Maximus") => 340.0,
        (false, "The Appian Way") => 160.0,
        (false, "The Hanging Gardens" | "The Sphinx") => 130.0,
        (false, "Piraeus") => 120.0,
        (false, "The Temple of Artemis") => 110.0,
        _ => 0.0,
    }
}

fn science_focus(
    g: &GameState,
    child: &GameState,
    action: &Action,
    player: usize,
    economy: bool,
) -> f64 {
    let before = science_symbols(g, player);
    let after = science_symbols(child, player);
    let before_count = before.into_iter().filter(|&x| x).count();
    let new_symbols = after
        .iter()
        .zip(before)
        .filter(|(after, before)| **after && !*before)
        .count();
    let pair_delta = child.cities[player].claimed_science_pairs.len() as i32
        - g.cities[player].claimed_science_pairs.len() as i32;
    let mut score = new_symbols as f64 * (700.0 + 140.0 * before_count as f64);
    if pair_delta != 0 {
        let law = data::progress_id("Law");
        let law_live = g.available_progress_tokens.contains(&law)
            && !g.cities[player].progress_tokens.contains(&law);
        score += pair_delta as f64 * if law_live { 900.0 } else { 260.0 };
    }
    if let Some(slot) = action.slot {
        let card = data::card(g.tableau.slots[slot].card_id);
        if card.color == CardColor::Green {
            if action.use_ == ActionUse::ConstructBuilding {
                score += 100.0;
                if card.age == 1 && card.science.is_some_and(|s| !before[s as usize]) {
                    score += 1_100.0;
                }
            } else if matches!(
                action.use_,
                ActionUse::DiscardForCoins | ActionUse::ConstructWonder
            ) {
                score -= if card.age == 1 { 900.0 } else { 350.0 };
            }
        }
    }
    if let Some(wid) = action.wonder {
        let name = data::wonder(wid).name;
        score += wonder_value(name, true);
        if name == "The Great Library" && progress_is_off_board(g, "Law") {
            score += 280.0;
        }
        if name == "The Mausoleum" {
            score += 80.0
                * g.discard_pile
                    .iter()
                    .filter(|&&cid| data::card(cid).color == CardColor::Green)
                    .count() as f64;
        }
        if economy && matches!(name, "Piraeus" | "The Great Lighthouse") {
            score += 130.0;
        }
    }
    if action.use_ == ActionUse::ResolvePendingChoice {
        let name = choice_name(g, action);
        score += match name {
            "Law" => 1_200.0,
            "Urbanism" | "Agriculture" => 180.0,
            "Economy" => 100.0,
            _ => 0.0,
        };
    }
    let support =
        resource_support_value(child, player, true) - resource_support_value(g, player, true);
    let coins = (child.cities[player].coins - g.cities[player].coins) as f64;
    score + if economy { 55.0 } else { 8.0 } * support + if economy { 1.8 } else { 0.4 } * coins
}

fn military_pressure(position: i32) -> f64 {
    25.0 * position as f64 + 8.0 * (position * position.abs()) as f64
}

fn military_focus(
    g: &GameState,
    child: &GameState,
    action: &Action,
    player: usize,
    economy: bool,
) -> f64 {
    let relative = |state: &GameState| {
        if player == 0 {
            state.conflict_position
        } else {
            -state.conflict_position
        }
    };
    let mut score = 3.0 * (military_pressure(relative(child)) - military_pressure(relative(g)));
    if let Some(slot) = action.slot {
        let card = data::card(g.tableau.slots[slot].card_id);
        if card.color == CardColor::Red {
            let amount = if card.age == 1 { 650.0 } else { 300.0 };
            score += if action.use_ == ActionUse::ConstructBuilding {
                amount
            } else if matches!(
                action.use_,
                ActionUse::DiscardForCoins | ActionUse::ConstructWonder
            ) {
                -amount
            } else {
                0.0
            };
        }
    }
    if let Some(wid) = action.wonder {
        let name = data::wonder(wid).name;
        score += wonder_value(name, false);
        if economy && matches!(name, "Piraeus" | "The Great Lighthouse") {
            score += 90.0;
        }
    }
    if action.use_ == ActionUse::ResolvePendingChoice {
        score += match choice_name(g, action) {
            "Strategy" => 1_200.0,
            "Urbanism" | "Agriculture" => 170.0,
            "Economy" => 100.0,
            _ => 0.0,
        };
    }
    let opponent = 1 - player;
    score += 8.0 * (g.cities[opponent].coins - child.cities[opponent].coins).max(0) as f64;
    let support =
        resource_support_value(child, player, false) - resource_support_value(g, player, false);
    let coins = (child.cities[player].coins - g.cities[player].coins) as f64;
    score + if economy { 55.0 } else { 8.0 } * support + if economy { 1.6 } else { 0.3 } * coins
}

fn rush_score(g: &GameState, action: &Action, kind: BotKind) -> f64 {
    let player = actor(g);
    let mut child = g.clone();
    child.apply_action(action);
    let science = matches!(kind, BotKind::ScienceAggressive | BotKind::ScienceEconomy);
    if child.phase == Phase::Complete {
        return if child.winner == Some(player) {
            if child.victory_type
                == Some(if science {
                    VictoryType::Scientific
                } else {
                    VictoryType::Military
                })
            {
                3_000_000.0
            } else {
                2_000_000.0
            }
        } else if child.winner.is_none() {
            0.0
        } else {
            -3_000_000.0
        };
    }
    let economy = matches!(kind, BotKind::ScienceEconomy | BotKind::MilitaryEconomy);
    let mut score = 0.02 * evaluate_state(&child, player);
    score += if science {
        science_focus(g, &child, action, player, economy)
    } else {
        military_focus(g, &child, action, player, economy)
    };
    if actor(&child) == player {
        score += 25.0;
    }
    if action.use_ == ActionUse::ChooseNextStartPlayer && action.starting_player == Some(player) {
        score += 15.0;
    }
    score
}

pub fn select_action(g: &GameState, kind: BotKind, rng: &mut Rng, exploration: f64) -> usize {
    let player = actor(g);
    let mut scored: Vec<(f64, _, Action)> = g
        .legal_actions()
        .into_iter()
        .map(|action| {
            let score = if kind == BotKind::Greedy {
                let mut child = g.clone();
                child.apply_action(&action);
                evaluate_state(&child, player)
            } else {
                rush_score(g, &action, kind)
            };
            (score, action_key(g, &action), action)
        })
        .collect();
    scored.sort_by(|a, b| {
        b.0.partial_cmp(&a.0)
            .unwrap_or(Ordering::Equal)
            .then_with(|| b.1.cmp(&a.1))
    });
    let selected = if kind != BotKind::Greedy && exploration > 0.0 && rng.next_float() < exploration
    {
        let limit = scored.len().min(3);
        &scored[rng.randrange(limit as u64) as usize].2
    } else {
        &scored[0].2
    };
    encode_action(g, selected)
}
